File: mqtt-camera/mqtt-cam.py
#
# This integration is towards MQTT in Home-Assistant and can easily
# be configured to provide both images streamed unfiltered or diff-filtered.
#
# Author: Joakim Eriksson, joakim.eriksson@ri.se
#

import paho.mqtt.client as mqttClient
import threading, time, yaml
import numpy as np, sys, time
import cv2
import images_pb2

def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("Connected to broker:", rc)
    else:
        print("Connection failed: ", rc)

class MQTTCamera:

    def __init__(self, mqttBroker, camera, topic="ha/camera/mqtt"):
        self.show = True
        self.broker = mqttBroker
        self.camera = camera
        self.topic = topic
        self.cap = cv2.VideoCapture(camera)
        # First frame is average...
        ret, self.avgframe = self.cap.read()
        self.client = mqttClient.Client("Python-CV-YOLO3")
        self.client.on_connect = on_connect
        self.client.connect(mqttBroker)
        self.client.loop_start()


    def diff_filter(self, frame, avgframe):
        subframe = cv2.subtract(frame, avgframe)
        grayscaled = cv2.cvtColor(subframe, cv2.COLOR_BGR2GRAY)
        retval2,th1 = cv2.threshold(grayscaled,35,255,cv2.THRESH_BINARY)
        avgframe = cv2.addWeighted(frame, 0.1, avgframe, 0.9, 0.0)

        if self.show:
            cv2.imshow('Treshold diff', th1)

        th1 = th1 / 255
        w, h = th1.shape
        sum = cv2.sumElems(th1)[0]/(w*h)
        return avgframe, sum

    def publish_image(self, frame):
        if self.client == None:
            return
        self.client.publish(self.topic, frame)
    
    def publish_image_pb(self, w, h, id, frame):
        if self.client == None:
            return
        img = images_pb2.Image()
        img.width = w
        img.height = h
        img.id = id
        img.imgdata = frame
        frame_pb = img.SerializeToString()
        print("Public as ", self.topic + "_pb")
        self.client.publish(self.topic + "_pb", frame_pb)
        

    def camera_loop(self):
        fc = 0
        while(True):
            # Capture frame-by-frame
            ret, frame = self.cap.read()

            self.avgframe, sum = self.diff_filter(frame, self.avgframe)
            if sum > 0.01:
                print("Publishing image diff:", sum)
                #self.publish_image(cv2.imencode('.png', frame)[1].tostring())
                h, w = frame.shape[:2]
                self.publish_image_pb(w, h, "hej", cv2.imencode('.png', frame)[1].tostring())

            # Our operations on the frame come here
            #frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
            # Display the resulting frame
            if self.show:
                fc = fc + 1

                if fc % 1 == 0:
                    cv2.imshow('frame',frame)
                    if cv2.waitKey(1) & 0xFF == ord('q'):
                        break
        self.cap.release()

video = 1
if len(sys.argv) > 1:
    video = sys.argv[1]

mqCam = MQTTCamera("localhost", video, topic="ha/camera/mqtt")
mqCam.show = False
mqCam.camera_loop()

# When everything done, release the capture
cv2.destroyAllWindows()

File: mqtt-camera/mqtt-cam.go
// Basic MQTT Camera source - for home assistant.
// Needs Paho MQTT client and opencv for GO (gocv) installed
// Author: Joakim Eriksson, joakim.eriksson@ri.se

package main

import (
	"fmt"
	"time"

	MQTT "github.com/eclipse/paho.mqtt.golang"
	mqtt "github.com/eclipse/paho.mqtt.golang"
	proto "github.com/golang/protobuf/proto"
	"gocv.io/x/gocv"

	pb "./images.pb"
)

type MqttCam struct {
	Host   string
	Topic  string
	Camera string
	Opts   *MQTT.ClientOptions
	Avg    gocv.Mat
}

func (mq *MqttCam) show() {
	fmt.Println(mq.Host, " with topic:", mq.Topic)
}

//define a function for the default message handler
var f MQTT.MessageHandler = func(client MQTT.Client, msg MQTT.Message) {
	fmt.Printf("TOPIC: %s\n", msg.Topic())
	fmt.Printf("MSG: %s\n", msg.Payload())
}

func main() {
	mcam := MqttCam{
		Host:   "localhost:1883",
		Topic:  "ha/camera/mqtt",
		Camera: "0",
	}

	//create a ClientOptions struct setting the broker address, clientid, turn
	//off trace output and set the default message handler
	mcam.Opts = MQTT.NewClientOptions().AddBroker(mcam.Host)
	mcam.Opts.SetClientID("go-simple")
	mcam.Opts.SetDefaultPublishHandler(f)

	mcam.show()

	c := mqtt.NewClient(mcam.Opts)
	c.Connect()

	//window := gocv.NewWindow("Hello")
	//w2 := gocv.NewWindow("Hello - Avg.")

	webcam, _ := gocv.OpenVideoCapture(mcam.Camera)

	mcam.Avg = gocv.NewMat()
	img := gocv.NewMat()
	sub := gocv.NewMat()
	gray := gocv.NewMat()
	th1 := gocv.NewMat()

	time.Sleep(3 * time.Second)

	webcam.Read(&mcam.Avg)

	// The camera loop
	for {
		webcam.Read(&img)
		gocv.Subtract(img, mcam.Avg, &sub)
		gocv.CvtColor(sub, &gray, gocv.ColorBGRToGray)
		gocv.Threshold(gray, &th1, 35, 255, gocv.ThresholdBinary)
		gocv.AddWeighted(img, 0.1, mcam.Avg, 0.9, 0.0, &mcam.Avg)

		th1.DivideFloat(255.0)
		h := float64(th1.Size()[0])
		w := float64(th1.Size()[1])
		diff := th1.Sum().Val1 / (w * h)
		if diff > 0.01 {
			fmt.Println("Sum", diff)
			imgdata, _ := gocv.IMEncode(".png", img)
			//fmt.Println("IMG:", imgdata, error)
			fmt.Println("Publish image on ", mcam.Topic)
			c.Publish(mcam.Topic, 0, false, string(imgdata))

			test := &pb.Image{
				Width:   int32(w),
				Height:  int32(h),
				Id:      "hej",
				Imgdata: imgdata,
			}
			data, err := proto.Marshal(test)
			fmt.Println("Publish pb image.", err)
			c.Publish(mcam.Topic+"_pb", 0, false, data)
		}
		//window.IMShow(img)
		//w2.IMShow(mcam.Avg)
		//window.WaitKey(1)
	}
	c.Disconnect(250)
}
